opendevin/controller/agent_controller.py

```diff
@@ -113,140 +128,117 @@ def add_history(self, action: Action, observation: Observation):
-                break
-
-            if self._task_state == TaskState.FINISHED:
-                logger.info('Task finished by agent')
-                await self.reset_task()
-                break
-            elif self._task_state == TaskState.STOPPED:
-                logger.info('Task stopped by user')
-                await self.reset_task()
-                break
-            elif self._task_state == TaskState.PAUSED:
-                logger.info('Task paused')
-                self._cur_step = i + 1
-                await self.notify_task_state_changed()
+                await self.set_agent_state_to(AgentState.STOPPED)
                 break
 
             if self._is_stuck():
                 logger.info('Loop detected, stopping task')
-                observation = AgentErrorObservation(
+                await self.add_error_to_history(
                     'I got stuck into a loop, the task has stopped.'
                 )
-                await self._run_callbacks(observation)
-                await self.set_task_state_to(TaskState.STOPPED)
+                await self.set_agent_state_to(AgentState.STOPPED)
                 break
+            await asyncio.sleep(
+                0.001
+            )  # Give back control for a tick, so other async stuff can run
 
     async def setup_task(self, task: str, inputs: dict = {}):
         """Sets up the agent controller with a task."""
-        self._task_state = TaskState.RUNNING
-        await self.notify_task_state_changed()
+        await self.set_agent_state_to(AgentState.INIT)
         self.state = State(Plan(task))
         self.state.inputs = inputs
 
-    async def start(self, task: str):
-        """Starts the agent controller with a task.
-        If task already run before, it will continue from the last step.
-        """
-        await self.setup_task(task)
-        await self._run()
-
-    async def resume(self):
-        if self.state is None:
-            raise ValueError('No task to resume')
-
-        self._task_state = TaskState.RUNNING
-        await self.notify_task_state_changed()
-
-        await self._run()
+    async def on_event(self, event: Event):
+        if isinstance(event, ChangeAgentStateAction):
+            if event.agent_state == AgentState.INIT:
+                await self.set_agent_state_to(AgentState.INIT)
+            elif event.agent_state == AgentState.RUNNING:
+                await self.set_agent_state_to(AgentState.RUNNING)
+            elif event.agent_state == AgentState.PAUSED:
+                await self.set_agent_state_to(AgentState.PAUSED)
+            elif event.agent_state == AgentState.STOPPED:
+                await self.set_agent_state_to(AgentState.STOPPED)
+            elif event.agent_state == AgentState.FINISHED:
+                await self.set_agent_state_to(AgentState.FINISHED)
+            else:
+                logger.warning(f'Unknown agent state: {event.agent_state}')
```

Comment:
it seems there're several other states in AgentState, like: ERROR/INIT/AWAITING_USER_INPUT, maybe `logger.warning(f'Unexpected agent state: {event.agent_state}')` is better?

Reply:
Good call!